### ISSUE TITLE
crash: panicked at 'assertion failed: `(left == right)` inside chalk

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-analyzer/rust-analyzer#6956]
````
git clone https://github.com/rust-random/rand
cd rand
git checkout 98a1aaf28e7bc272c12796d96401c177bf2205e4
RUST_BACKTRACE=full rust-analyzer analysis-stats .  --with-deps
````
````
Database loaded 1.14s, 80mi
Crates in this dir: 88
Total modules found: 1064
Total declarations: 33684
Total functions: 21855
Item Collection: 6.41s, 31155mi
21519/21855 98% processing: boxed::pinthread 'main' panicked at 'assertion failed: `(left == right)`
  left: `2`,
 right: `1`', /home/matthias/.cargo/registry/src/github.com-1ecc6299db9ec823/chalk-ir-0.43.0/src/lib.rs:2093:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f74583445702e2e27ec4415376f2c540a83d7ded/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/f74583445702e2e27ec4415376f2c540a83d7ded/library/core/src/panicking.rs:92:14
   2: chalk_ir::Binders<T>::substitute
   3: chalk_solve::clauses::builtin_traits::sized::add_sized_program_clauses
   4: chalk_solve::clauses::builder::ClauseBuilder<I>::push_binders
   5: chalk_solve::clauses::program_clauses_for_goal
   6: <chalk_recursive::recursive::Solver<I> as chalk_recursive::solve::SolveDatabase<I>>::solve_goal
   7: <chalk_recursive::recursive::RecursiveSolver<I> as chalk_solve::solve::Solver<I>>::solve_limited
   8: hir_ty::traits::trait_solve_query
   9: salsa::runtime::Runtime::execute_query_implementation
  10: salsa::derived::slot::Slot<Q,MP>::read_upgrade
  11: salsa::derived::slot::Slot<Q,MP>::read
  12: <salsa::derived::DerivedStorage<Q,MP> as salsa::plumbing::QueryStorageOps<Q>>::try_fetch
  13: <DB as hir_ty::db::HirDatabase>::trait_solve::__shim
  14: <DB as hir_ty::db::HirDatabase>::trait_solve
  15: hir_ty::infer::InferenceContext::resolve_ty_as_possible
  16: hir_ty::infer::expr::<impl hir_ty::infer::InferenceContext>::infer_method_call
  17: hir_ty::infer::expr::<impl hir_ty::infer::InferenceContext>::infer_expr_inner
  18: hir_ty::infer::expr::<impl hir_ty::infer::InferenceContext>::infer_expr_coerce
  19: hir_ty::infer::expr::<impl hir_ty::infer::InferenceContext>::infer_block
  20: hir_ty::infer::expr::<impl hir_ty::infer::InferenceContext>::infer_expr_inner
  21: hir_ty::infer::expr::<impl hir_ty::infer::InferenceContext>::infer_expr_coerce
  22: hir_ty::infer::infer_query
  23: salsa::runtime::Runtime::execute_query_implementation
  24: salsa::derived::slot::Slot<Q,MP>::read_upgrade
  25: salsa::derived::slot::Slot<Q,MP>::read
  26: <salsa::derived::DerivedStorage<Q,MP> as salsa::plumbing::QueryStorageOps<Q>>::try_fetch
  27: <DB as hir_ty::db::HirDatabase>::infer_query::__shim
  28: hir_ty::db::infer_wait
  29: rust_analyzer::cli::analysis_stats::AnalysisStatsCmd::run
  30: rust_analyzer::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
````
ra @ 02fb3cb542c676ffcd627d5ac5092aa3a5d71d30

**Comment by tweksteen** [comment]
Same issue here, using the tip of the `release` branch (commit 60c501f).

**Comment by tweksteen** [comment]
I initially found this issue while trying to process the [`boxed::pin` definition](https://github.com/rust-lang/rust/blob/master/library/alloc/src/boxed.rs#L239) in `boxed.rs`
```
processing: boxed::pin (/aosp/master/prebuilts/rust/linux-x86/1.49.0/src/stdlibs/library/alloc/src/boxed.rs 7433..7707)
thread 'main' panicked at 'assertion failed: `(left == right)`
[...]
```

This is also triggered when parsing nightly `core` from `stdlibs`. To reproduce:
```
$ rust-analyzer --version
rust-analyzer fde4a86
$ pwd
~/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core
$ CHALK_DEBUG=1 rust-analyzer analysis-stats .
> solving UCanonical { canonical: Canonical { value: InEnvironment { environment: Env([]), goal: Implemented(Box<[?0 := ^0.0]>: Sized) }, binders: [U0 with kind integer type] }, universes: 1 }

thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `2`,
```
With `CHALK_PRINT=1`, this is the program passed to chalk from rust-analyzer:
```
chalk program:
#[upstream]
#[non_enumerable]
#[object_safe]
trait Allocator {}
#[upstream]
#[non_enumerable]
#[object_safe]
#[lang(sized)]
trait Sized {}
#[upstream]
struct Box<_1_0, _1_1>
where
  _1_1: Allocator
{}
``` 

In this case, the panic is triggered in `test_clone_from` from [core/tests/clone.rs](https://github.com/rust-lang/rust/blob/3aa325221041bc4aba3ffc637b7a2cd475617aad/library/core/tests/clone.rs#L10):

```
fn test_clone_from() {
    let a = box 5;
    let mut b = box 10;
    b.clone_from(&a);
    assert_eq!(*b, 5);
}
```

In both cases, this seems to be related to the [`box` keyword](https://doc.rust-lang.org/beta/unstable-book/language-features/box-syntax.html).

1. @matklad Could you review the chalk program above?
2. @matklad Would you accept a pull request for an integration test (before `release`) to run `analysis-stats` on `stdlibs`?

Thanks

**Comment by matklad** [comment]
cc @flodiebold for 1

for 2, extending this logic seems appropriate: https://github.com/rust-analyzer/rust-analyzer/blob/master/.github/workflows/release.yaml#L111-L113

**Comment by flodiebold** [comment]
The problem is that our handling of the `box` keyword constructs a `Box` with just one type parameter, but now the type has two parameters:
https://github.com/rust-analyzer/rust-analyzer/blob/cc53252e8718540cf392c1cd40951ef685bd52a0/crates/hir_ty/src/infer/expr.rs#L491-L498

We should check the number of type parameters there. I think we can probably just fill it up with type variables. It might be worth checking whether there are other places where we make assumptions on the number of parameters of a (non-lang item) standard library type.

**Comment by matthiaskrgr** [comment]
Was this supposed to be closed?

**Comment by bjorn3** [comment]
It should have been fixed by #7386.
